dnssd: support service subtypes

rpigott opened the pull request:
A service subtype is used for selective enumeration of services. [1]

[1] https://datatracker.ietf.org/doc/html/rfc6763#section-7.1

rpigott commented:
Fixes #16069 I guess

poettering commented:
this is marked as draft still. any reason for that. (we usually don't review drafts, hence if you want this reviewed, please drop the "draft" status).

i reviewed this anyway now, looks pretty much fine. maybe a man page update and things should be good.

rpigott commented:
Yeah this was intentionally left as a draft. #22532 is a more valuable feature to me and I wasn't sure if this would block it (no immediate interaction, but it may have implications for the varlink api?). If necessary I don't mind updating it after #22532 is merged if changes are needed. I will undraft it and you can decide, in case you think this should go in first.

rpigott commented:
Added manpage bit.

mrc0mmand commented:
> > TEST-75-RESOLVED: [ 2729.281326] ns1.unsigned.test systemd-resolved[515]: varlink-30: Return parameters for method reply ResolveService() didn't pass validation on field 'canonical', ignoring: No anode
> 
> This method doesn't exist at this point in history tho? Does this need a rebase or smth?

Yup, please rebase, the error should be fixed by #31369.

evverx commented on src/resolve/resolved-dnssd.c:
```diff
@@ -253,6 +262,16 @@ int dnssd_update_rrs(DnssdService *s) {
                 goto oom;
 
+        s->sub_ptr_rr = dns_resource_record_new_full(DNS_CLASS_IN, DNS_TYPE_PTR,
+                                                 selective_name);
```
Looks like services without subtypes crash resolved:
```sh
Using system hostname 'H'.                                                                                                                                    
Assertion 'name' failed at src/resolve/resolved-dns-rr.c:24, function dns_resource_key_new(). Aborting.
...
#0  0x00007ffff56b0884 in __pthread_kill_implementation () from /lib64/libc.so.6
#1  0x00007ffff565fafe in raise () from /lib64/libc.so.6
#2  0x00007ffff564887f in abort () from /lib64/libc.so.6
#3  0x00007ffff71074f1 in log_assert_failed (text=0x8a8720 <str> "name", file=0x8ac684 "src/resolve/resolved-dns-rr.c", line=24, 
    func=0x8a8780 <__func__.dns_resource_key_new> "dns_resource_key_new") at ../src/basic/log.c:968
#4  0x00000000007cc62d in dns_resource_key_new (class=1, type=12, name=0x0) at ../src/resolve/resolved-dns-rr.c:24
#5  0x00000000007d2feb in dns_resource_record_new_full (class=1, type=12, name=0x0) at ../src/resolve/resolved-dns-rr.c:393
#6  0x0000000000692e89 in dnssd_update_rrs (s=0x60800000cba0) at ../src/resolve/resolved-dnssd.c:265
#7  0x00000000006909dd in dnssd_service_load (manager=0x619000000f80, filename=0x6030001541b0 "/etc/systemd/dnssd/qotd.dnssd")
    at ../src/resolve/resolved-dnssd.c:131
#8  0x000000000068d667 in dnssd_load (manager=0x619000000f80) at ../src/resolve/resolved-dnssd.c:203
#9  0x00000000006f01c6 in manager_new (ret=0x7ffff2609020) at ../src/resolve/resolved-manager.c:629
#10 0x000000000076c056 in run (argc=1, argv=0x7fffffffe4b8) at ../src/resolve/resolved.c:72
#11 0x000000000076ab0e in main (argc=1, argv=0x7fffffffe4b8) at ../src/resolve/resolved.c:99
```
Services with subtypes seem to trigger a memory leak:
```sh
==140469==ERROR: LeakSanitizer: detected memory leaks
Direct leak of 112 byte(s) in 1 object(s) allocated from:
    #0 0x4c2f12 in malloc (/home/vagrant/systemd-asan/build/systemd-resolved+0x4c2f12) (BuildId: 72669fc0ab2b83b1baac44fc30ddde9a38aecd33)
    #1 0x7cebc1 in malloc_multiply /home/vagrant/systemd-asan/build/../src/basic/alloc-util.h:119:16
    #2 0x7d2c61 in dns_resource_record_new /home/vagrant/systemd-asan/build/../src/resolve/resolved-dns-rr.c:375:14
    #3 0x7d3062 in dns_resource_record_new_full /home/vagrant/systemd-asan/build/../src/resolve/resolved-dns-rr.c:397:16
    #4 0x692e88 in dnssd_update_rrs /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:265:25
    #5 0x6909dc in dnssd_service_load /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:131:13
    #6 0x68d666 in dnssd_load /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:203:21
    #7 0x6f01c5 in manager_new /home/vagrant/systemd-asan/build/../src/resolve/resolved-manager.c:629:13
    #8 0x76c055 in run /home/vagrant/systemd-asan/build/../src/resolve/resolved.c:72:13
    #9 0x76ab0d in main /home/vagrant/systemd-asan/build/../src/resolve/resolved.c:99:1
    #10 0x7f3856e49b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #11 0x7f3856e49c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #12 0x4282c4 in _start (/home/vagrant/systemd-asan/build/systemd-resolved+0x4282c4) (BuildId: 72669fc0ab2b83b1baac44fc30ddde9a38aecd33)
Indirect leak of 47 byte(s) in 1 object(s) allocated from:
    #0 0x4c304c in __interceptor_calloc (/home/vagrant/systemd-asan/build/systemd-resolved+0x4c304c) (BuildId: 72669fc0ab2b83b1baac44fc30ddde9a38aecd33)
    #1 0x7cc6ba in dns_resource_key_new /home/vagrant/systemd-asan/build/../src/resolve/resolved-dns-rr.c:27:13
    #2 0x7d2fea in dns_resource_record_new_full /home/vagrant/systemd-asan/build/../src/resolve/resolved-dns-rr.c:393:15
    #3 0x692e88 in dnssd_update_rrs /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:265:25
    #4 0x6909dc in dnssd_service_load /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:131:13
    #5 0x68d666 in dnssd_load /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:203:21
    #6 0x6f01c5 in manager_new /home/vagrant/systemd-asan/build/../src/resolve/resolved-manager.c:629:13
    #7 0x76c055 in run /home/vagrant/systemd-asan/build/../src/resolve/resolved.c:72:13
    #8 0x76ab0d in main /home/vagrant/systemd-asan/build/../src/resolve/resolved.c:99:1
    #9 0x7f3856e49b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #10 0x7f3856e49c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #11 0x4282c4 in _start (/home/vagrant/systemd-asan/build/systemd-resolved+0x4282c4) (BuildId: 72669fc0ab2b83b1baac44fc30ddde9a38aecd33)
Indirect leak of 19 byte(s) in 1 object(s) allocated from:
    #0 0x4accf1 in strdup (/home/vagrant/systemd-asan/build/systemd-resolved+0x4accf1) (BuildId: 72669fc0ab2b83b1baac44fc30ddde9a38aecd33)
    #1 0x6931e4 in dnssd_update_rrs /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:271:35
    #2 0x6909dc in dnssd_service_load /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:131:13
    #3 0x68d666 in dnssd_load /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:203:21
    #4 0x6f01c5 in manager_new /home/vagrant/systemd-asan/build/../src/resolve/resolved-manager.c:629:13
    #5 0x76c055 in run /home/vagrant/systemd-asan/build/../src/resolve/resolved.c:72:13
    #6 0x76ab0d in main /home/vagrant/systemd-asan/build/../src/resolve/resolved.c:99:1
    #7 0x7f3856e49b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #8 0x7f3856e49c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #9 0x4282c4 in _start (/home/vagrant/systemd-asan/build/systemd-resolved+0x4282c4) (BuildId: 72669fc0ab2b83b1baac44fc30ddde9a38aecd33)
SUMMARY: AddressSanitizer: 178 byte(s) leaked in 3 allocation(s).
```

poettering replied:
we break lines differently btw. i.e. line width is 109ch. if you break, break after the `(` of the function call, and indent to twice the usual indentation (i.e. 16 spaces)

evverx commented:
subtypes don't seem to advertised anywhere:
```
12:12:14.442667 IP H.mdns > mdns.mcast.net.mdns: 0 [2q] [3n] ANY (QM)? H.local. ANY (QM)? H._qotd._tcp.local. (100)
12:12:14.743995 IP H.mdns > mdns.mcast.net.mdns: 0 [2q] [3n] ANY (QM)? H.local. ANY (QM)? H._qotd._tcp.local. (100)
12:12:15.192266 IP H.mdns > mdns.mcast.net.mdns: 0 [2q] [3n] ANY (QM)? H.local. ANY (QM)? H._qotd._tcp.local. (100)
12:12:15.446584 IP H.mdns > mdns.mcast.net.mdns: 0*- [0q] 6/0/0 (Cache flush) PTR H.local., (Cache flush) SRV H.local.:123 0 0, (Cache flush) TXT "", (Cache flush) A 192.168.56.4, PTR H._qotd._tcp.local., PTR _qotd._tcp.local. (178)
12:12:16.687643 IP H.mdns > mdns.mcast.net.mdns: 0*- [0q] 6/0/0 (Cache flush) A 192.168.56.4, (Cache flush) PTR H.local., (Cache flush) SRV H.local.:123 0 0, (Cache flush) TXT "", PTR _qotd._tcp.local., PTR H._qotd._tcp.local. (178)
```
My guess would be that it's because `dns_scope_add_dnssd_services` hasn't been updated. `bus_dnssd_method_unregister` should probably be updated too to make it possible to at least deregister those services via D-Bus properly. (In theory `RegisterService` can accept subtypes along with types via the third parameter. It should be backward-compatible as far as I can tell)

rpigott commented:
Thanks for the comments, but TBH my use case for this feature still has not materialized, so I don't really care to fix it up atm.

I intended to use it advertising lan cache servers for pacman, but the feature is only supported in the still unreleased pacman 7.0. I guess that's another reason I let this one languish as a draft. When I get the chance I'ma fix up the DNR patches because I might want to actually use those.

evverx commented:
Got it. It's just that I thought that this PR was on its way to the main branch so I decided to chime in :-)

> I don't really care to fix it up atm

Got it. I think it came at the right time though because the browsing/resolving API stuff is on its way and I'm not sure it takes subtypes into account at all. Maybe "type" can be overloaded in the sense that it could accept "_wat._sub._qotd._tcp" and "_qotd._tcp" but last time I checked it didn't work without instance names and there are no instance names with subtypes.

poettering commented on src/resolve/resolved-conf.c:
```diff
@@ -299,6 +299,42 @@ int config_parse_dnssd_service_type(
+                return 0;
+        }
+
+        r = free_and_strdup(&s->subtype, rvalue);
```
return free_and_strdup_warn(…);

poettering commented on src/resolve/resolved-dnssd-gperf.gperf:
```diff
@@ -18,6 +18,7 @@ struct ConfigPerfItem;
 %%
 Service.Name,     config_parse_dnssd_service_name, 0,                 0
 Service.Type,     config_parse_dnssd_service_type, 0,                 0
+Service.SubType,  config_parse_dnssd_service_subtype, 0,              0
```
please just rebreak the table so tha tthe columns fit again

poettering commented:
looks quite good actually, just minor comments.

rpigott commented:
So, they did eventually release this feature in pacman 6.1. Think I addressed the review comments, and added in some mysteriously missing parts, like where we actually add the newly made records.

> It's possible to pass subtypes there in the "domainName" argument:

We need some special handling for these anyway, since we need to advertise a PTR from the selective name with subtype to the usual name without a subtype.

evverx commented:
> We need some special handling for these anyway, since we need to advertise a PTR from the selective name with subtype to the usual name without a subtype.

Agreed. I think what I was trying to say was that it should be possible to overload the "type" argument of the "RegisterService" method without changing its signature to keep it backward-compatible. It could extract the subtype part from types passed to it.

Another option would of course be a separate method named, say, "RegisterServiceWithSubtype" with an additional argument.

rpigott commented:
Ah, I understand now. No opinion either way really, but yeah that makes sense.

evverx commented on src/resolve/resolved-conf.c:
```diff
@@ -299,6 +299,37 @@ int config_parse_dnssd_service_type(
+        }
+
+        return free_and_strdup_warn(&s->subtype, rvalue);
+}
```
I'm not sure if it's related to this PR but with `SubType=` I got
```sh
=================================================================
==241773==ERROR: LeakSanitizer: detected memory leaks
Direct leak of 11 byte(s) in 1 object(s) allocated from:
    #0 0x7fd82dcd4e91 in __interceptor_strdup (/usr/lib64/clang/16/lib/libclang_rt.asan-x86_64.so+0xd4e91) (BuildId: 2a39bb4aa6d13ff866a4a47997cefdbe31a0350f)
    #1 0x7fd82ce3ef56 in free_and_strdup /home/vagrant/systemd-asan/build/../src/basic/string-util.c:1092:21
    #2 0x43d444 in config_parse_dnssd_service_type /home/vagrant/systemd-asan/build/../src/resolve/resolved-conf.c:295:13
    #3 0x7fd82c484fea in next_assignment /home/vagrant/systemd-asan/build/../src/shared/conf-parser.c:156:24
    #4 0x7fd82c469243 in parse_line /home/vagrant/systemd-asan/build/../src/shared/conf-parser.c:262:16
    #5 0x7fd82c4662d6 in config_parse /home/vagrant/systemd-asan/build/../src/shared/conf-parser.c:405:21
    #6 0x7fd82c46ba00 in config_parse_many_files /home/vagrant/systemd-asan/build/../src/shared/conf-parser.c:561:21
    #7 0x7fd82c46a132 in config_parse_many /home/vagrant/systemd-asan/build/../src/shared/conf-parser.c:623:13
    #8 0x5a339b in dnssd_service_load /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:92:13
    #9 0x5a1596 in dnssd_load /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:204:21
    #10 0x6042c5 in manager_new /home/vagrant/systemd-asan/build/../src/resolve/resolved-manager.c:629:13
    #11 0x683c55 in run /home/vagrant/systemd-asan/build/../src/resolve/resolved.c:72:13
    #12 0x68270d in main /home/vagrant/systemd-asan/build/../src/resolve/resolved.c:99:1
    #13 0x7fd82b249b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #14 0x7fd82b249c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #15 0x40d534 in _start (/home/vagrant/systemd-asan/build/systemd-resolved+0x40d534) (BuildId: 331b141e8a7288fe8d80eb125eff50797c791e17)
SUMMARY: AddressSanitizer: 11 byte(s) leaked in 1 allocation(s).
```

rpigott replied:
Pasta error.

evverx commented:
Looks like it works. `_http._sub._qotd._tcp.local.` was advertised by resolved and the mDNS responder on another machine saw it. The "Deregister" method deregistered the service too in the sense that goodbye packets were sent and `dns-sd` showed that:
```
dns-sd -B _qotd._tcp,_http
Browsing for _qotd._tcp,_http
DATE: ---Sat 16 Mar 2024---
 6:26:16.939  ...STARTING...
Timestamp     A/R    Flags  if Domain               Service Type         Instance Name
 6:28:09.800  Add        2  13 local.               _qotd._tcp.          H
 6:38:38.578  Rmv        0  13 local.               _qotd._tcp.          H
```

Interestingly it turns out resolved advertises all RRs with ttl=120 even though according to https://datatracker.ietf.org/doc/html/rfc6762#section-10 some RRs should come with ttl=4500. Known-answer suppression doesn't seem to work:
```
<Ether  dst=01:00:5e:00:00:fb src=4a:d7:05:2c:1d:64 type=IPv4 |<IP  version=4 ihl=5 tos=0x0 len=89 id=12913 flags= frag=0 ttl=255 proto=udp chksum=0x0 src=192.168.56.3 dst=224.0.0.251 |<UDP  sport=mdns dport=mdns len=69 chksum=0xd9fd |<DNS  id=0 qr=0 opcode=QUERY aa=0 tc=0 rd=0 ra=0 z=0 ad=0 cd=0 rcode=ok qdcount=1 ancount=1 nscount=0 arcount=0 qd=[<DNSQR  qname=b'_http._sub._qotd._tcp.local.' qtype=PTR qclass=IN |>] an=[<DNSRR  rrname=b'_http._sub._qotd._tcp.local.' type=PTR rclass=IN ttl=115 rdata=b'H._qotd._tcp.local.' |>] |>>>>
<Ether  dst=01:00:5e:00:00:fb src=08:00:27:0b:2f:0b type=IPv4 |<IP  version=4 ihl=5 tos=0x0 len=83 id=12564 flags= frag=0 ttl=255 proto=udp chksum=0xb0dd src=192.168.56.4 dst=224.0.0.251 |<UDP  sport=mdns dport=mdns len=63 chksum=0x70a |<DNS  id=0 qr=1 opcode=QUERY aa=1 tc=0 rd=0 ra=0 z=0 ad=0 cd=0 rcode=ok qdcount=0 ancount=1 nscount=0 arcount=0 an=[<DNSRR  rrname=b'_http._sub._qotd._tcp.local.' type=PTR rclass=IN ttl=120 rdata=b'H._qotd._tcp.local.' |>] |>>>>
```
but it has nothing to do with this PR.

evverx commented:
Looks like it violates the RFC (a bit) in the sense that `resolved` advertises RRs pointing `_services._dns-sd._udp.local` to `_http._sub._qotd._tcp.local.`:
```sh
-bash-5.2# dig @192.168.124.17 -p 5353 _services._dns-sd._udp.local ANY +notcp +noall +ans
_services._dns-sd._udp.local. 120 IN    PTR     _http._sub._qotd._tcp.local.
_services._dns-sd._udp.local. 120 IN    PTR     _qotd._tcp.local.
```
but according to https://datatracker.ietf.org/doc/html/rfc6763#section-9 the rdata should be two labels plus a domain:
> A DNS query for
   PTR records with the name "_services._dns-sd._udp.<Domain>" yields a
   set of PTR records, where the rdata of each PTR record is the two-
   label <Service> name, plus the same domain, e.g.,
   "_http._tcp.<Domain>".

I've just checked and `mDNSResponder` doesn't advertise RRs like that. It just adds PTR RRs pointing subtypes to service instance names.
